File: problems/prob2 - Even Fibonacci numbers/sln.go
package main

import "fmt"

var fib_memo = make(map[int]int)

func main() {
	fib_memo[0] = 1
	fib_memo[1] = 1
	term_count := 2
	fib_value := fib(term_count)
	sum := fib_value
	for sum < 4000000 {
		term_count += 1
		fib_value := fib(term_count);
		if fib_value % 2 == 0 {
			sum += fib_value;
		}
	}
	fmt.Println(sum)
}

func fib(n int) int {
	memo_value, contains_key := fib_memo[n]
	if contains_key {
		return memo_value
	}

	fib_memo[n] = fib(n-1) + fib(n-2)
	return fib_memo[n]
}

File: problems/prob2 - Even Fibonacci numbers/sln.java
import java.util.Hashtable;


class Main {
    static Hashtable<Integer, Integer> fib_memo = new Hashtable<Integer, Integer>();

    public static void main(String[] args) {
        // base case
        fib_memo.put(0, 1);
        fib_memo.put(1, 1);
        Integer term_count = 2;
        Integer fib_value = fib(term_count);
        Integer sum = fib_value;
        while(fib_value < 4000000) {
            term_count += 1;
            fib_value = fib(term_count);
            if(fib_value % 2 == 0) {
                sum += fib_value;
            }
        }

        System.out.println(sum);
    }

    public static Integer fib(Integer n) {
        // check if n is memoized
        if(fib_memo.containsKey(n)) {
            return fib_memo.get(n);
        }

        // otherwise calculate it
        Integer fib_result = fib(n-1) + fib(n-2);
        fib_memo.put(n, fib_result);
        return fib_result;
    }
}
